Treat explicit instantiations as full uses

jru opened the pull request:
Modify IsFowardDecl() to check that the decl is not an explicit template instantiation (declaration or definition) This check is also in its own function, IsExplicitInstantiation(), for convenience.

jru commented:
Well, in this case the header file doesn't play any role, that's why I didn't add any. 

The result of the test previous to the change suggests the instantiation definition for the *ExplicitlyDeclaredAndDefined* template to get removed: 
```
tests/cxx/explicit_template_instantiation.cc should add these lines:

tests/cxx/explicit_template_instantiation.cc should remove these lines:
- class ExplicitlyDeclaredAndDefined;  // lines 32-32

The full include-list for tests/cxx/explicit_template_instantiation.cc:
```
Honestly, I haven't analyzed completely why the third declaration and other two end up being treated as full uses (I am guessing something in the semantics changes when one follows the other), but in any case we don't want them to be confused with forward declarations, so I made sure to keep them.

jru commented on iwyu_ast_util.cc:
```diff
@@ -942,12 +942,21 @@ bool IsFriendDecl(const Decl* decl) {
 }
 
+bool IsExplicitInstantiation(const clang::CXXRecordDecl* decl) {
+  const auto kind = decl->getTemplateSpecializationKind();
```
The name of the getter contains the enum type, *TemplateSpecializationKind*. That's why I thought it would be clear. I can repeat myself, no problem ;)

kimgr commented:
I think a more typical layout/use of explicit instantiations is to have the template definition and explicit instantiation decl in a header file, and the explicit instantiation defn in a .cpp file.

So rather than a minimized test case to provoke this bug, we could put up some representative code. Something like:

```c++
// explicit_template_instantiation-template.h

// Base template
template<class T>
class Template {};

// Explicit instantiation declaration.
extern template class Template<int>;


// explicit_template_instantiation-short.h

// Another explicit instantiation declaration.
extern template class Template<short>;

// explicit_template_instantiation.cc
#include "explicit_template_instantiation-template.h"
#include "explicit_template_instantiation-short.h"

// Define explicit instantiation
template class Template<int>;

// Some uses
void f() {
    // Use the base template
    Template<bool> t0;

    // Use the locally-defined explicit instantiation
    Template<int> t1;

    // Use the declared but not defined explicit instantiation
    // (I'm guessing this should type-check, but not link?)
    // Should make sure the -short.h header is not removed
    Template<short> t2;
}
```

I left out the `// IWYU` assertions in the interest of brevity (and because I'm not entirely sure what might happen). Overall this seems like a more realistic test case to me.

jru commented:
I played with it a bit more and I found some other cases where IWYU is not doing the right thing. 
Specifically, explicit instantiation declarations of templated functions seem to be ignored. 

This needs more work than previously thought..

kimgr commented:
I think we should leave function templates as a known issue to make some headway with what we have here.

jru commented:
Yes, I don't think they can be tackled the same way since they are not visited directly, unless `shouldVisitTemplateInstantiations()` returns true, and that breaks currently. I'll create an issue to address this later.

Class and Var templates are almost working, but I got stuck for a while. I think I could finalize this tomorrow.

jru commented:
I finally managed to finalize this, for class templates so far. 

Not sure if we can add something else to the test case. What you mentioned in your example above about verifying that a header file with a explicit instantiation declaration is included is not really possible, since such declaration is not necessary for the usage of template itself.

kimgr commented:
I think the tests would benefit from separation like I suggested earlier. Maybe you need to be explicit with  `--check_also` to get proper output from headers.

As for the `short` case, that's really the most interesting to me -- see the review feedback.

jru commented:
> We included a header saying there's a singular instantiation for short and if that include is removed, we'll implicitly instantiate it, generating more code. Or if the base template has no definition, we'll fail to compile.

I guess it makes sense. But, it'll need some work :)
So, what should be expected behaviour? Which uses of that template should make the decl stay? 

For example, looking at the 7 cases below. Should we keep the include for all of them, or only the uses of the same specialization? And only full uses, or also fwd decl uses? What about implicit uses?

```cpp
// file template.h
template<class>
class Template {};

// file template-decl-short.h
extern template class Template<short>;

//file main.cc
#include "template.h"
#include "template-decl-short.h"

template <class T> class FullUseArg { T t; }
template <class T> class FwdDeclUseArg { T* t; }
template <class T = Template<short>> class TemplateAsDefaultFull { T t; };
template <class T = Template<short>> class TemplateAsDefaultFwd { T* t; };

Template<short> s; // 1
template Template<short>; // 2
FullUseArg<Template<short>> fu; // 3
FwdDeclUseArg<Template<short>> fd; // 4
TemplateAsDefaultFull<> td; // 5
TemplateAsDefaultFwd<> td; // 6
Template<int> s; // 7

```

I would go for 1-3 and 5, but maybe that misses other cases. For example, if this is itself an include and there are only forward decl uses of the template, but it is then fully used somewhere else.

Also note, if this is implemented the include 'template-decl-short.h' will be required in all compilation units that use that specialization. Do we want that?

kimgr commented:
> Also note, if this is implemented the include 'template-decl-short.h' will be required in all compilation
> units that use that specialization. Do we want that?

I think so... Or you mean that it could be transitively satisfied, so the header doesn't need to be included everywhere? From a strict IWYU perspective it would be nice to explicitly include that specialization everywhere it's used because it makes the code resilient to refactors -- if the specialization include is removed from an incidental dependency, our code will silently change meaning (using the base template instead).

I find it hard to reason about the seven cases when they're all in the same file, but I'll give it a shot...

```
template <class T = Template<short>> class TemplateAsDefaultFull { T t; };
template <class T = Template<short>> class TemplateAsDefaultFwd { T* t; };
```

I'm not sure this is 100% consistent, but it seems to me IWYU's general rule is that the file that spells out the type should include the type. So template argument defaults should be attributed to the file declaring the template. I can't play this out in my head for the templates above, so not sure how forward-decl vs full uses should behave.

`Template<short> s; // 1`

This seems like it should obviously require `template-decl-short.h`.

`template Template<short>; // 2`

Explicit template instantiation definition of `Template<short>`, right? That should require `template-decl-short.h`, otherwise it will create a reusable instantiation of the wrong template.

`FullUseArg<Template<short>> fu; // 3`

`Template<short>` is spelled out here and `fu` constructor will need to see the full type, so `template-decl-short.h` required.

`FwdDeclUseArg<Template<short>> fd; // 4`

`Template<short>` is spelled out but constructing `fd` does not need the full type, so a forward-decl or explicit instantiation declaration of `Template<short>` should do here.

```
TemplateAsDefaultFull<> td; // 5
TemplateAsDefaultFwd<> td; // 6
```

These are tricky.

The general rule says we shouldn't need `Template<short>` because it's not spelled out. But it is required for constructing `td`. I think this is one of the author-intent rules -- `Template<short>` should be attributed to the file declaring `TemplateAsDefaultFull` unless it has a forward-declare of `Template<short>` to signal that users should be responsible. And I think it should always require the complete type of the default template arguments, because I can't get my head around the analysis of what might happen when the templates are instantiated.

`Template<int> s; // 7`

No use of `Template<short>`, so this doesn't count toward that header.

kimgr commented:
I also think it's hard to reason about this, because I keep confusing "explicit instantiation" with "explicit specialization".

Specializations, whether partial or total, alter the behavior of the program if visible to the compilation unit. 

Explicit instantiations, on the other hand, seem to be more of a build-time optimization: we can use them to help the compiler do separate compilation of templates.

The challenge in IWYU is that we want to preserve the behavior of the program *and* not worsen build times. So it would seem we can never remove a specialization or explicit instantiation unless we can prove that they're unused.

I'm pretty sure this is not helping :), I just needed to type it out to understand it better. I'm really out of my depth when it comes to template handling in IWYU.

jru commented:
I just added some work-in-progress that already satisifies 1 and 2. For the rest, there is more work to do.

Also something I noticed is that IWYU tries very hard to normalize decls and their locations. :) For usual template instantiations that's not a problem (as we usually want the definition), but for explicit instantiation decls, I really had to fight it so that it would leave the decl I pass in to *ReportDeclUse* alone. 

I am not happy with the result so far, so I would like to gather more input before I continue with the other cases.

jru commented:
FYI, I am still working on this. It turned out to be much harder than I thought, but I think I start to see the light at the end of the tunnel :)

jru commented:
I finally managed to finish this. There are a few more commits that in the beginning, so let me explain briefly (more details in the commits themselves):

*These are the three main changes to make explicit template instantiations work:*
- Treat explicit template instantiations as full uses
- Require explicit inst. decls to use main template
- Use the definition to check if decl is provided by template

*These contain some refactoring work that simplifies the implementation:*
- Make 'comment' a default argument in Report* functions
- Refactor CanIgnoreType for clarity and reuse

*Supporting change to be able to report specific redecls:*
- Introduce an extra use flag UF_TargetRedecl

*Unrelated bugfix discovered during the implementation:*
- Fix GetLocation for templated functions

*Also unrelated to my changes. Seems to be failing after a new warning from clang:*
- Silence warning -Wsizeof-pointer-div in test

kimgr commented:
Very cool! I haven't had a chance to look at the body of the template instantiation logic, but it strikes me that many of the refactors would be nice on their own. Can you separate them out and post them as individual PRs? The problem with huge PRs like this is that they're both hard to review and blocked until all commits are ready to go in, and much of this looks useful even before trying to fix the explicit instantiation problems.

I'll actually start off by cherry-picking the `-Wnosizeof-pointer` commit, to get master green again. I'll come back with what I think is a logical order of changes afterwards.

kimgr commented:
Is `Fix GetLocation for templated functions` as free-standing as it looks? It's nice when you can remove code to fix a broken test :) Is there a test that should be written to capture the new behavior?

If it doesn't depend on anything else, I think this would be nice to get on master first.

kimgr commented:
`Introduce an extra use flag UF_TargetRedecl`: I feel there's more than this going on in this commit.

It seems suspicious that `UF_FunctionDfn` and `UF_TargetRedecl` control the same behavior, but I intended for use-flags to say something about the context in which the use was detected, so if they describe different contexts, they should be different. That said, I'm not sure what `UF_TargetRedecl` says...?

Overall, I feel like if we could break this commit into its constituent parts, maybe there are fragments that could go in earlier and solve smaller problems than the elusive explicit template instantiation :)

There's a lot of spurious reformatting in this commit (and maybe others) -- please tread lightly; it's so much harder to review (and blame in the future) when lines are reformatted as part of a functional change.

jru commented:
Hey,

I extracted the two refactor commits here: #607. I also rebased this branch on top of it to be sure there are no conflicts. Once it is merged I'll rebase again on top of master.


> Introduce an extra use flag UF_TargetRedecl

You were right about the whitespace, I messed it up somehow. Should be fixed now! Also made some small changes to ReportFullSymbolUse to hopefully make it clearer.

> That said, I'm not sure what UF_TargetRedecl says...?

The main idea here was to pass extra information to the reporting logic that says "please, treat this decl as-is". I went back and forth trying out different ideas and this seemed like the cleanest solution. Although, not perfect by any means :)

Right now, the flag is not used inside OneUse for anything else, but I can imagine this might turn out to be useful for other tings. In fact, during the implementation I had to use the UF_TargetRedecl to reimplement `DeclCanBeForwardDeclared` as a member of OneUse, but later on turned out not to be necessary for this case, so I removed it to make the patch smaller.

The existing UF_FunctionDfn is used to implement `OneUse::is_function_being_defined` so I cannot really mix the two. That said, it might not be the right thing to do to add another flag for this purpose. I am open for ideas :)

In general, it seems to me the reporting side of IWYU `iwyu_output.*` shouldn't know anything about Clang or the AST and instead work exclusively based on the interface provided by OneUse (tiny as of right now), so that it becomes easier to extend and maintain in the long run.

jru commented:
> Is Fix GetLocation for templated functions as free-standing as it looks? 

Yes, it is. I dropped it from the branch and will try to think of a test for it.

kimgr commented:
> The existing UF_FunctionDfn is used to implement OneUse::is_function_being_defined so I cannot
> really mix the two. That said, it might not be the right thing to do to add another flag for this purpose. I 
> am open for ideas :)

Ah, I see. No, I think a new flag is probably warranted.

> In general, it seems to me the reporting side of IWYU iwyu_output.* shouldn't know anything about 
> Clang or the AST and instead work exclusively based on the interface provided by OneUse (tiny as of 
> right now), so that it becomes easier to extend and maintain in the long run.

I've actually been considering going in the opposite direction -- capture and expose parts of the AST and expose to `iwyu_output.*`.

Some things are hard to classify correctly when traversing the AST because they require cross-examination of multiple structures (can't think of an example now, but `UF_FunctionDfn` came about to solve one of those problems). The idea of use-flags is to capture some information about the use-location and forward it to later analysis where we have more context. I've even played with the idea of refactoring `UseFlags` into a `UseContext` class to forward e.g. AST fragments into `iwyu_output` (which I think is necessary to solve #105, for example).

So I'm not sure I share the concern that `iwyu_output.*` should not know anything about prior phases. And I think the new use-flag should say something about the context in which the use occurred, not what the earlier phase thinks the later phase should do with it.

jru commented:
Hi, refactors are in and branch rebased.

> I've actually been considering going in the opposite direction -- capture and expose parts of the AST and expose to iwyu_output.*.

Yes, that might work too. The main point I was trying to make is that the visiting side is lower level, and as such, it doesn't have the complete picture. However, it does have access to all of the details about the AST. The analysis (reporting/iwyu_output) side, on the other hand, is much higher level, with much more context, but has lost much of the detail from the lower phase.

The main problem, as I see it, is that this separation of concerns is not well defined or maintained in the code. For example, the visiting side tries to make high-level decisions about what things to report, whereas the analysis will very often overrule those decisions at a later point after doing some low-level visiting of certain nodes. 

I think this interplay between the two should be represented with better abstractions. Maybe the analysis phase needs to use the AST directly after all, but I can imagine that a higher-level model of the AST where we expose what we need and hide the rest, could be more benefitial in the long run.

kimgr commented:
> The main point I was trying to make is that the visiting side is lower level, and as such, it doesn't have 
> the complete picture. However, it does have access to all of the details about the AST. The analysis 
> (reporting/iwyu_output) side, on the other hand, is much higher level, with much more context, but 
> has lost much of the detail from the lower phase.

Exactly, that's what I was trying to say too.

> The main problem, as I see it, is that this separation of concerns is not well defined or maintained in 
> the code.

Agreed!

> For example, the visiting side tries to make high-level decisions about what things to report, whereas 
> the analysis will very often overrule those decisions at a later point after doing some low-level visiting 
> of certain nodes.
> I think this interplay between the two should be represented with better abstractions. Maybe the 
> analysis phase needs to use the AST directly after all, but I can imagine that a higher-level model
> of the AST where we expose what we need and hide the rest, could be more benefitial in the long
> run.

Yes, I think there might be a phase missing to sort out stuff like author intent. That is;

- AST visiting to collect "raw" uses
- Refinement phase to take stuff like author intent into account and move responsibilities around
- Analysis phase to play out the IWYU rules

There are abstractions missing, for sure, and some amount of purity where phases don't make too many "forward" assumptions. I.e. visiting the AST should be mostly oblivious to whether a use will count for something (except possibly for `CanIgnore*` type things to avoid collecting too much data).

And the Clang AST data structures are mostly wonderful to work with, so it would be nice to use them as a lingua franca throughout the pipeline (with some additional decoration like parent maps, etc) instead of inventing our own representations.

I'm regularly missing an include graph abstraction, where it's possible to ask about the "path" between two files (e.g. between the main file and `GetFileEntry(some_symbol)`), and I'm sure there are others.

jru commented:
Yep, all good ideas! :) Sadly, it is not obvious where to start with it all while not breaking anything..

Btw, would you like me to rework the UF_TargetRedecl commit? Or is it clear enough?

jru commented on tests/cxx/explicit_template_instantiation-template.h:
```diff
@@ -0,0 +1,14 @@
+//=== explicit_template_instantiation_template.h - test input file for iwyu ===//
```
Done

jru commented on tests/cxx/explicit_template_instantiation-template.h:
```diff
@@ -0,0 +1,14 @@
+
+template<typename T> struct TemplateClass {};
+
+#endif  //INCLUDE_WHAT_YOU_USE_TESTS_CXX_EXPLICIT_TEMPLATE_INSTANTIATION_TEMPLATE_H_
```
Done

jru commented on tests/cxx/explicit_template_instantiation-template_direct.h:
```diff
@@ -0,0 +1,14 @@
+
+#include "explicit_template_instantiation-template.h"
+
+#endif  //INCLUDE_WHAT_YOU_USE_TESTS_CXX_EXPLICIT_TEMPLATE_INSTANTIATION_TEMPLATE_DIRECT_H_
```
Done

kimgr commented:
> Btw, would you like me to rework the UF_TargetRedecl commit? Or is it clear enough?

Yes, I think it would be nice if it followed existing form and described the context of the use, rather than telling the next phase what to do.

jru commented on iwyu_ast_util.cc:
```diff
@@ -942,12 +942,28 @@ bool IsFriendDecl(const Decl* decl) {
   return decl->getFriendObjectKind() != Decl::FOK_None;
 }
 
+clang::TemplateSpecializationKind GetTemplateSpecializationKind(
```
Done

jru commented:
> Yes, I think it would be nice if it followed existing form and described the context of the use

Right. The thing is that it is not really a flag for the use, but for the decl associated to it. I could name it *UF_ExplicitInstantiation* and explain that the use targets one.

Another option is to create another set of `DeclFlags`, but it seems like overdoing it a little.

kimgr commented:
`UF_ExplicitInstantiation` works for me!

jru commented:
Done. All other comments were addressed too.
If you are good with it I'll rebase the fixups and reword the commit.

kimgr commented:
I'd rather review the changes commit-by-commit, so please rewrite the commit history to look the way you want to ship it. I started looking at the total diff, and overall it looks really nice!

jru commented:
It is ready.

kimgr commented on tests/cxx/explicit_instantiation.cc:
```diff
@@ -0,0 +1,48 @@
+template struct Template<double>;
+
+
+// The explicit instantiation of a specialization needs a declaration
```
"... *only* needs a declaration..."?

kimgr commented on iwyu_ast_util.cc:
```diff
@@ -942,12 +950,20 @@ bool IsFriendDecl(const Decl* decl) {
+
 bool IsForwardDecl(const NamedDecl* decl) {
-  if (const auto* record_decl = dyn_cast<RecordDecl>(decl)) {
+  if (const CXXRecordDecl* record_decl = DynCastFrom(decl)) {
```
Hmm, `DynCastFrom` predates `auto` and I think it's clearer to use LLVM's `dyn_cast` and friends with `auto` than leaning on this hack.

To be perfectly honest, I think it's slightly easier to read the `DynCastFrom` line, but I had to spend a considerable amount of time understand how it worked the first time I saw it. `auto` is well-understood.

kimgr commented on iwyu_ast_util.cc:
```diff
@@ -155,6 +155,14 @@ void DumpASTNode(llvm::raw_ostream& ostream, const ASTNode* node) {
 
+clang::TemplateSpecializationKind GetTemplateSpecializationKind(
+    const Decl* decl) {
+  if (const clang::CXXRecordDecl* record = DynCastFrom(decl)) {
```
See below -- I think we should prefer `const auto* x = dyn_cast<T>(y)` over `const T* x = DynCastFrom(y)`.

jru replied:
Alright. I actually thought we settled for `DynCasFrom`, but yeah, I also think `dyn_cast` is a better choice.

kimgr commented on iwyu.cc:
```diff
@@ -3732,11 +3732,20 @@ class IwyuAstConsumer
   //    template<> struct Foo<int> { ... };
   // we don't want iwyu to recommend removing the 'forward declare' of Foo.
+  //
+  // Additionally, this type of decl is also used to represent template
```
I remember reading the Josuttis/Vandevoorde book on C++ templates about a decade ago, and they consistently used "specialization" instead of "instantiation", that is, they were the same thing in the standard text back then.

I think "explicit instantiation" is the primary context in which "instantiation" is used, even now. So maybe this comment should say "... is also used to represent *explicit* template instantiations...". Everything else is a specialization, not an instantiation.

Also, the "Additionally ... also" pair looks like a double-take. I think you can remove "also".

jru replied:
Yes, I meant to write *explicit* instantiations. Thanks!

Regarding the usage of the terms, I must admit I find the whole thing quite confusing:
Looking at how the [C++ standard](from http://eel.is/c++draft/temp.spec) uses them it seems to me that *instantiation* is the process of substitution and generation, and also the resulting types/methods (ClassTemplateSpecializationDecl in clang), whereas *specializations* are the template definitions before the process of instantiation (the ClassTemplateDecl in clang)

But now looking at clang, `ClassTemplateSpecializationDecl` *is* an instantiated `CXXRecordDecl` whereas `ClassTemplateDecl` *references* the uninstantiated `CXXRecordDecl`, which seems a bit backward to me :)

kimgr commented on iwyu.cc:
```diff
@@ -109,6 +109,7 @@
 #include "iwyu_location_util.h"
 #include "iwyu_output.h"
 #include "iwyu_path_util.h"
+#include "iwyu_use_flags.h"
```
Commit message needs some updating after this use-flag was repurposed.

kimgr commented on iwyu.cc:
```diff
@@ -1632,29 +1633,32 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (const UsingShadowDecl* shadow_decl = DynCastFrom(used_decl))
       target_decl = shadow_decl->getTargetDecl();
-    
+
```
Let's do this cleanup as a separate no-functional-change commit.

kimgr commented on iwyu.cc:
```diff
@@ -1668,12 +1672,10 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
-    if (const UsingDecl* using_decl
-        = GetUsingDeclarationOf(used_decl, 
-              GetDeclContext(current_ast_node()))) {
+    if (const UsingDecl* using_decl = GetUsingDeclarationOf(
```
Spurious reformat

kimgr commented on iwyu_output.cc:
```diff
@@ -578,14 +578,19 @@ void IwyuFileInfo::ReportFullSymbolUse(SourceLocation use_loc,
+
+    // However, if we're defining the function or we are targeting an explicit
+    // instantiation, we want to use it as-is and not try to canonicalize at all.
+    if ((flags & (UF_FunctionDfn | UF_ExplicitInstantiation)) != 0) {
```
There should be one or more testcases enabled by this change... Any chance you can get them into the same commit?

jru replied:
> There should be one or more testcases enabled by this change
You mean, new ones I should write? Without this change the tests wouldn't pass, so it is technically covered.

kimgr replied:
Sorry, I didn't see this before commenting further. It seems to me this change came about for a reason, and it would be nice if that reason was expressed as tests and committed together with the change.

But maybe the tests were failing before this change? Or it makes more sense to commit them with later changes? I think this looks good as-is if it turns out weird with tests together with this.

kimgr commented on iwyu_output.cc:
```diff
@@ -578,14 +578,19 @@ void IwyuFileInfo::ReportFullSymbolUse(SourceLocation use_loc,
+
+    // However, if we're defining the function or we are targeting an explicit
+    // instantiation, we want to use it as-is and not try to canonicalize at all.
+    if ((flags & (UF_FunctionDfn | UF_ExplicitInstantiation)) != 0) {
```
I wonder if it would be possible to turn this into an if/else block instead, something like:
```c++
SourceLocation decl_loc;

if (flags & (UF_FunctionDfn | UF_ExplicitInstantiation) == 0) {
    // Rationale here...
    decl = GetDefinitionAsWritten(decl);
    decl_loc = GetLocation(report_decl);
} else {
    // ... and here
    decl_loc = report_decl->getLocation();
}
```
That way it looks more balanced (even if it's annoying with the `decl_loc` declaration outside the block.

kimgr commented on iwyu.cc:
```diff
@@ -2548,12 +2548,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
```
Commit message:  "when there is a explicit instantiation definition visible" -> "... *an* explicit..."

kimgr commented on iwyu.cc:
```diff
@@ -2548,12 +2548,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
     const NamedDecl* decl = TypeToDeclAsWritten(type);
 
+    const SourceLocation current_loc = CurrentLoc();
```
No need to outline this into a variable, just put the inline `CurrentLoc` calls back

kimgr commented on iwyu.cc:
```diff
@@ -3110,9 +3113,53 @@ class InstantiatedTemplateVisitor
     ReportTypeUse(caller_loc(), actual_type);
+
+    // Report also all previous explicit instantiations (declarations and
+    // defintions) as uses of the caller's location.
```
Typo: defintions

kimgr commented on iwyu.cc:
```diff
@@ -3110,9 +3113,53 @@ class InstantiatedTemplateVisitor
     // template's caller.
     ReportTypeUse(caller_loc(), actual_type);
+
+    // Report also all previous explicit instantiations (declarations and
```
Report also -> Also report

kimgr commented on iwyu.cc:
```diff
@@ -3110,9 +3113,53 @@ class InstantiatedTemplateVisitor
   }
 
+  bool VisitTemplateSpecializationType(
+      clang::TemplateSpecializationType* type) {
```
`TemplateSpecializationType` has been pulled into the namespace with `using` at the top of the file. If you remove the namespace specifier, the declaration fits on one line.

kimgr commented on iwyu.cc:
```diff
@@ -3110,9 +3113,53 @@ class InstantiatedTemplateVisitor
+    if (CanIgnoreCurrentASTNode())
+      return true;
+
+    CHECK_(current_ast_node()->GetAs<clang::TemplateSpecializationType>() ==
```
Same here, lose `clang::`

kimgr commented on iwyu.cc:
```diff
@@ -3110,9 +3113,53 @@ class InstantiatedTemplateVisitor
+  }
+
+  void ReportExplicitInstantiations(const Type* type) {
+    const ClassTemplateSpecializationDecl* decl =
```
Prefer `auto` + `dyn_cast<>`

kimgr commented on tests/cxx/explicit_instantiation2-template_helpers.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef INCLUDE_WHAT_YOU_USE_TESTS_CXX_EXPLICIT_INSTANTIATION2_TEMPLATE_HELPERS_H_
+#define INCLUDE_WHAT_YOU_USE_TESTS_CXX_EXPLICIT_INSTANTIATION2_TEMPLATE_HELPERS_H_
+
+template <class T>
```
Is this a re-declaration of the other `Template` in explicit_instantiation-template.h? In that case, maybe use `struct` instead of `class`, so they're consistent -- some compilers warn when decls differ only by class/struct keyword.

If not, please choose a unique-r name.

kimgr commented on tests/cxx/explicit_instantiation2.cc:
```diff
@@ -0,0 +1,99 @@
+// required, or it does not provide an explicit instantiation:
+// 6: Fwd-decl use in a template, provided as a default parameter.
+// 9: Implicit instantiation of Template<int>
+//10: Specialization of Template<T>
```
The alignment looks nice, but I prefer to always have a space after `//`.

kimgr commented on tests/cxx/explicit_instantiation2.cc:
```diff
@@ -0,0 +1,99 @@
+// These tests verify that a dependent explicit instantiation declaration is
+// always required when the main template is also needed, but not otherwise.
+//
+// Positive scenarios:
```
Maybe these comments would have better locality next to the respective test case if possible?

kimgr commented on tests/cxx/explicit_instantiation2.cc:
```diff
@@ -0,0 +1,99 @@
+
+// IWYU: Template is...*explicit_instantiation-template.h
+// IWYU: Template is...*template_short.h.*for explicit instantiation
+Template<short> t1a;
```
Ah, now I see you've used naming to refer back to the list... Maybe better to use comments consistently, like `// 2` below?

kimgr commented on tests/cxx/explicit_instantiation2.cc:
```diff
@@ -29,6 +29,8 @@
 // 6: Fwd-decl use in a template, provided as a default parameter.
 // 9: Implicit instantiation of Template<int>
 //10: Specialization of Template<T>
+//11: Explicit instantiation with a dependent instantiation declaration, but
```
Prefer space after `//` to alignment of numbered list.

jru commented:
Thanks for the kind words! 
I think I addressed all the comments, except [this one](https://github.com/include-what-you-use/include-what-you-use/pull/578#discussion_r236791395). Could you clarify what you meant?

jru commented on iwyu.cc:
```diff
@@ -1649,12 +1652,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (CanIgnoreDecl(target_decl))
       return;
 
+    const UseFlags use_flags =
```
Well, I think is good practice to prevent accidental modifications, especially in long functions. It also helps document the intent.

kimgr commented:
\o/
